File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/3. Advanced Maths/python/04. Prime factorization using Sieve.py
"""QUESTION:
Prime-factorize N using sieve-style marking.

Approach:
Maintain a boolean array for primality. For each prime i up to sqrt(original N), repeatedly divide N while storing i in the factor list, and mark multiples of i as composite. If the remainder exceeds 1, append it.

Time Complexity: O(N log log N)
Space Complexity: O(N)
"""
from typing import List

def find_prime_factors(n: int) -> List[int]:
    if n <= 1:
        return []
    limit = n
    prime = [True] * (limit + 1)
    ans: List[int] = []
    i = 2
    while i * i <= limit:
        if prime[i]:
            while n % i == 0:
                ans.append(i)
                n //= i
            j = i * i
            while j <= limit:
                prime[j] = False
                j += i
        i += 1
    if n > 1:
        ans.append(n)
    return ans

File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/3. Advanced Maths/javascript/04. Prime factorization using Sieve.js
/**
QUESTION:
Prime-factorize N using sieve logic.

Approach:
Maintain a boolean array for primality. For each prime i up to sqrt(N), repeatedly divide N while pushing i into the result, and mark multiples of i as composite. If the remainder is > 1, append it.

Time Complexity: O(N log log N)
Space Complexity: O(N)
*/

function findPrimeFactors(n) {
    if (n <= 1) return [];
    const original = n;
    const prime = new Array(original + 1).fill(true);
    const ans = [];
    for (let i = 2; i * i <= original; i++) {
        if (prime[i]) {
            while (n % i === 0) {
                ans.push(i);
                n = Math.floor(n / i);
            }
            for (let j = i * i; j <= original; j += i) {
                prime[j] = false;
            }
        }
    }
    if (n > 1) ans.push(n);
    return ans;
}

module.exports = { findPrimeFactors };
